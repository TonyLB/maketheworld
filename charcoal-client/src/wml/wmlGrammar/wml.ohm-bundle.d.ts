// AUTOGENERATED FILE
// This file was generated from wml.ohm by `ohm generateBundles`.

import {
  ActionDict,
  Grammar,
  IterationNode,
  Node,
  NonterminalNode,
  Semantics,
  TerminalNode
} from 'ohm-js';

export interface WorldMarkupLangageActionDict<T> extends ActionDict<T> {
  WMLFileContents?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  CharacterExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  CharacterLegalContents?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  AssetExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  AssetLegalContents?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  ImportExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  ImportLegalContents?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  UseExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  DependencyExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  VariableExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  ComputedExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  ActionExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  ConditionExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  ConditionLegalContents?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  LayerExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  LayerLegalContents?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  ExitExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  LinkExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  RoomExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  RoomContents?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  LiteralNameExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  NameExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  LiteralValueTag?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  ValueTag?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  TextContents?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  TagExpression?: (this: NonterminalNode, arg0: NonterminalNode, arg1: IterationNode, arg2: NonterminalNode) => T;
  string?: (this: NonterminalNode, arg0: IterationNode) => T;
  stringText?: (this: NonterminalNode, arg0: IterationNode) => T;
  legalKey?: (this: NonterminalNode, arg0: IterationNode) => T;
  spaceCompressor?: (this: NonterminalNode, arg0: IterationNode) => T;
  TagSelfClosing?: (this: NonterminalNode, arg0: TerminalNode, arg1: Node, arg2: IterationNode, arg3: TerminalNode) => T;
  TagOpen?: (this: NonterminalNode, arg0: TerminalNode, arg1: Node, arg2: IterationNode, arg3: TerminalNode) => T;
  TagArgument?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  TagLiteralArgument?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  TagProgramArgument?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  tagBooleanArgument?: (this: NonterminalNode, arg0: Node, arg1: NonterminalNode | TerminalNode) => T;
  tagArgumentQuoted?: (this: NonterminalNode, arg0: Node, arg1: TerminalNode, arg2: NonterminalNode) => T;
  TagArgumentKey?: (this: NonterminalNode, arg0: Node, arg1: TerminalNode, arg2: NonterminalNode, arg3: TerminalNode) => T;
  TagArgumentBracketed?: (this: NonterminalNode, arg0: Node, arg1: TerminalNode, arg2: IterationNode, arg3: TerminalNode) => T;
  tagArgValueQuoted?: (this: NonterminalNode, arg0: IterationNode, arg1: TerminalNode) => T;
  TagClose?: (this: NonterminalNode, arg0: TerminalNode, arg1: Node, arg2: TerminalNode) => T;
  none?: (this: NonterminalNode, arg0: TerminalNode) => T;
  EmbeddedJSExpression?: (this: NonterminalNode, arg0: TerminalNode, arg1: IterationNode, arg2: TerminalNode) => T;
  JSExpression?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  JSDoubleQuotedString?: (this: NonterminalNode, arg0: TerminalNode, arg1: IterationNode, arg2: TerminalNode) => T;
  JSNoDoubleQuote?: (this: NonterminalNode, arg0: NonterminalNode | TerminalNode) => T;
  JSSingleQuotedString?: (this: NonterminalNode, arg0: TerminalNode, arg1: IterationNode, arg2: TerminalNode) => T;
  JSNoSingleQuote?: (this: NonterminalNode, arg0: NonterminalNode | TerminalNode) => T;
  JSTemplateString?: (this: NonterminalNode, arg0: TerminalNode, arg1: IterationNode, arg2: TerminalNode) => T;
  JSTemplateEvaluation?: (this: NonterminalNode, arg0: TerminalNode, arg1: NonterminalNode, arg2: TerminalNode) => T;
  JSNoBackQuote?: (this: NonterminalNode, arg0: NonterminalNode | TerminalNode) => T;
  JSText?: (this: NonterminalNode, arg0: IterationNode) => T;
  space?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  sourceCharacter?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  whitespace_verticalTab?: (this: NonterminalNode, arg0: TerminalNode) => T;
  whitespace_formFeed?: (this: NonterminalNode, arg0: TerminalNode) => T;
  whitespace_noBreakSpace?: (this: NonterminalNode, arg0: TerminalNode) => T;
  whitespace_byteOrderMark?: (this: NonterminalNode, arg0: TerminalNode) => T;
  whitespace?: (this: NonterminalNode, arg0: NonterminalNode | TerminalNode) => T;
  unicodeSpaceSeparator?: (this: NonterminalNode, arg0: TerminalNode) => T;
  lineTerminator?: (this: NonterminalNode, arg0: TerminalNode) => T;
  lineTerminatorSequence?: (this: NonterminalNode, arg0: TerminalNode) => T;
  comment?: (this: NonterminalNode, arg0: NonterminalNode) => T;
  multiLineComment?: (this: NonterminalNode, arg0: TerminalNode, arg1: IterationNode, arg2: TerminalNode) => T;
  singleLineComment?: (this: NonterminalNode, arg0: TerminalNode, arg1: IterationNode) => T;
}

export interface WorldMarkupLangageSemantics extends Semantics {
  addOperation<T>(name: string, actionDict: WorldMarkupLangageActionDict<T>): this;
  extendOperation<T>(name: string, actionDict: WorldMarkupLangageActionDict<T>): this;
  addAttribute<T>(name: string, actionDict: WorldMarkupLangageActionDict<T>): this;
  extendAttribute<T>(name: string, actionDict: WorldMarkupLangageActionDict<T>): this;
}

export interface WorldMarkupLangageGrammar extends Grammar {
  createSemantics(): WorldMarkupLangageSemantics;
  extendSemantics(superSemantics: WorldMarkupLangageSemantics): WorldMarkupLangageSemantics;
}

declare const grammar: WorldMarkupLangageGrammar;
export default grammar;

