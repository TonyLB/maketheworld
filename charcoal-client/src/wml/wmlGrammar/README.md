The wml.ohm-bundle files are generated:  Only the wml.ohm file is original.

To regenerate (e.g. after changing the wml.ohm file), navigate to the src/wml directory, and
execute this command:  npx ohm generateBundles --withTypes 'wmlGrammar/*.ohm'