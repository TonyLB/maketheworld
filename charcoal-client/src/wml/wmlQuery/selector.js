import wmlGrammar from '../wmlGrammar/wml.ohm-bundle'
import { Interval } from 'ohm-js'

import wmlQueryGrammar from '../wmlGrammar/wmlQuery.ohm-bundle'

// interface WMLSelection {
//     source: Interval;
//     node: Node;
// }

export const wmlSelectorSemantics = wmlQueryGrammar.createSemantics()
    .addOperation("parse", {
        TagAncestry(tags) {
            return {
                selector: 'TagAncestry',
                tags: tags.parse()
            }
        },
        WMLLegalTag(value) {
            return this.sourceString
        },
        _iter(...nodes) {
            return nodes.map((node) => (node.parse()))
        }    
    })

const wmlQuerySemantics = wmlGrammar.createSemantics()
    .addOperation("toNode", {
        TagOpen(open, tag, props, close) {
            return {
                tag: tag.toNode(),
                props: Object.assign({}, ...(props.toNode() || {}))
            }
        },
        TagSelfClosing(open, tag, props, close) {
            return {
                tag: tag.toNode(),
                props: Object.assign({}, ...(props.toNode() || {})),
                contents: []
            }
        },
        TagExpression(open, contents, close) {
            return {
                ...open.toNode(),
                contents: contents.toNode()
            }
        },
        tagBooleanArgument(key, spacing) {
            return { [key.toNode()]: true }
        },
        tagArgumentQuoted(key, equal, value) {
            return { [key.toNode()]: value.sourceString.slice(0, -1) }
        },
        TagArgumentBracketed(key, equal, value, close) {
            return { [key.toNode()]: value.sourceString }

        },
        string(node) {
            return this.sourceString
        },
        _iter(...contents) {
            return contents.map((node) => (node.toNode()))
        },
        _terminal() {
            return this.sourceString
        }
    })
    .addOperation("search(selector)", {
        TagOpen(open, tag, props, close) {
            const { tags } = this.args.selector
            return {
                tagMatch: tags.length > 0 && (tag.sourceString === tags[0])
            }
        },
        TagExpression(open, contents, close) {
            const { tagMatch } = open.search(this.args.selector)
            if (tagMatch) {
                const { tags, selector } = this.args.selector
                if (tags.length > 1) {
                    const remainingTags = tags.slice(1)
                    return contents.search({ selector, tags: remainingTags })
                }
                else {
                    return [{
                        node: this.toNode(),
                        source: {
                            start: this.source.startIdx,
                            end: this.source.endIdx
                        },
                    }]
                }
            }
            else {
                return contents.search(this.args.selector)
            }
        },
        _iter(...contents) {
            return contents.reduce((previous, child) => ([...previous, ...child.search(this.args.selector)]), [])
        },
        _terminal() {
            return []
        }
        //
        // TODO: Create a first cut semantics engine that generates a tree of Tags with leafs of their inner text.
        // Figure out a way to return *both* the relevant and searchable semantic data of the tag, *and* the
        // Intervals into the source string, that will allow for future editing of the string.
        //
    })

export const wmlSelectorFactory = (schema) => (searchString) => {
    const match = wmlQueryGrammar.match(searchString)
    if (!match.succeeded()) {
        return []
    }
    const selector = wmlSelectorSemantics(match).parse()
    //
    // TODO: Pass the selector to the querySemantics above, in order to parse out the
    // specific nodes/ranges being looked at
    //
    return wmlQuerySemantics(schema).search(selector)
}
