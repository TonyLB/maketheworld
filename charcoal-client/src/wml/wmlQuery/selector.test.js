import wmlGrammar from '../wmlGrammar/wml.ohm-bundle.js'
import { wmlSelectorFactory } from './selector.js'

describe('wmlQuery selector', () => {

    const characterMatch = wmlGrammar.match(`
        <Character key=(TESS) fileName="Tess" player="TonyLB">
            <Name>Tess</Name>
            <Pronouns
                subject="she"
                object="her"
                possessive="her"
                adjective="hers"
                reflexive="herself"
            ></Pronouns>
            <FirstImpression>Frumpy Goth</FirstImpression>
            <OneCoolThing>Fuchsia eyes</OneCoolThing>
            <Outfit>A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.</Outfit>
        </Character>
    `)
    let characterQuery = wmlSelectorFactory(characterMatch)

    const assetMatch = wmlGrammar.match(`
        <Asset key=(BASE)>
            <Room key=(VORTEX) global>
                Test Render:
                <Link key=(123) to=(clockTower)>Clock Tower</Link>
                <Exit to=(Test)>test</Exit>
                <Exit from=(Test)>vortex</Exit>
            </Room>
            <Room key=(Test) />
            <Feature key=(clockTower)>
                Clocktower
                test
                on multiple lines
            </Feature>
        </Asset>
    `)
    let assetQuery = wmlSelectorFactory(assetMatch)

    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
        characterQuery = wmlSelectorFactory(characterMatch)
        assetQuery = wmlSelectorFactory(assetMatch)
    })

    it('should return empty on illegal selector', () => {
        expect(characterQuery('Fraggle Rock')).toEqual([])
    })

    it('should correctly select root node', () => {
        expect(characterQuery('Character')).toMatchSnapshot()
    })

    it('should select root node when passed empty string', () => {
        expect(characterQuery('')).toMatchSnapshot()
    })

    it('should correctly select leaf node', () => {
        expect(characterQuery('Name')).toMatchSnapshot()
    })

    it('should correctly select ancestor chain', () => {
        expect(characterQuery('Character Name')).toMatchSnapshot()
    })

    it('should select nothing on a nonmatching chain', () => {
        expect(characterQuery('Name Outfit')).toEqual([])
    })

    it('should return a list for multiple matches', () => {
        expect(assetQuery('Room')).toMatchSnapshot()
    })

    it('should correctly subset by property', () => {
        expect(assetQuery('Room[key="VORTEX"]')).toMatchSnapshot()
    })

    it('should properly chain complex predicates', () => {
        expect(assetQuery('Room[key="VORTEX"] Exit[to="Test"]')).toMatchSnapshot()
    })

    it('should properly select :first filter', () => {
        expect(assetQuery('Room Exit:first')).toMatchSnapshot()
    })
})