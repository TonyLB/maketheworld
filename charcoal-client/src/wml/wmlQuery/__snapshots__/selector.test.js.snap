// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wmlQuery selector should correctly select ancestor chain 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "end": 86,
        "start": 82,
        "type": "string",
        "value": "Tess",
      },
    ],
    "contentsEnd": 86,
    "contentsStart": 82,
    "end": 93,
    "props": Object {},
    "start": 76,
    "tag": "Name",
    "tagEnd": 81,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should correctly select leaf node 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "end": 86,
        "start": 82,
        "type": "string",
        "value": "Tess",
      },
    ],
    "contentsEnd": 86,
    "contentsStart": 82,
    "end": 93,
    "props": Object {},
    "start": 76,
    "tag": "Name",
    "tagEnd": 81,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should correctly select root node 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "contents": Array [
          Object {
            "end": 86,
            "start": 82,
            "type": "string",
            "value": "Tess",
          },
        ],
        "contentsEnd": 86,
        "contentsStart": 82,
        "end": 93,
        "props": Object {},
        "start": 76,
        "tag": "Name",
        "tagEnd": 81,
        "type": "tag",
      },
      Object {
        "contents": Array [],
        "contentsEnd": 290,
        "contentsStart": 290,
        "end": 301,
        "props": Object {
          "adjective": Object {
            "end": 240,
            "start": 224,
            "value": "hers",
            "valueEnd": 239,
            "valueStart": 235,
          },
          "object": Object {
            "end": 174,
            "start": 162,
            "value": "her",
            "valueEnd": 173,
            "valueStart": 170,
          },
          "possessive": Object {
            "end": 207,
            "start": 191,
            "value": "her",
            "valueEnd": 206,
            "valueStart": 203,
          },
          "reflexive": Object {
            "end": 276,
            "start": 257,
            "value": "herself",
            "valueEnd": 275,
            "valueStart": 268,
          },
          "subject": Object {
            "end": 145,
            "start": 132,
            "value": "she",
            "valueEnd": 144,
            "valueStart": 141,
          },
        },
        "start": 106,
        "tag": "Pronouns",
        "tagEnd": 115,
        "type": "tag",
      },
      Object {
        "contents": Array [
          Object {
            "end": 342,
            "start": 331,
            "type": "string",
            "value": "Frumpy Goth",
          },
        ],
        "contentsEnd": 342,
        "contentsStart": 331,
        "end": 360,
        "props": Object {},
        "start": 314,
        "tag": "FirstImpression",
        "tagEnd": 330,
        "type": "tag",
      },
      Object {
        "contents": Array [
          Object {
            "end": 399,
            "start": 387,
            "type": "string",
            "value": "Fuchsia eyes",
          },
        ],
        "contentsEnd": 399,
        "contentsStart": 387,
        "end": 414,
        "props": Object {},
        "start": 373,
        "tag": "OneCoolThing",
        "tagEnd": 386,
        "type": "tag",
      },
      Object {
        "contents": Array [
          Object {
            "end": 517,
            "start": 435,
            "type": "string",
            "value": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
          },
        ],
        "contentsEnd": 517,
        "contentsStart": 435,
        "end": 526,
        "props": Object {},
        "start": 427,
        "tag": "Outfit",
        "tagEnd": 434,
        "type": "tag",
      },
    ],
    "contentsEnd": 535,
    "contentsStart": 63,
    "end": 547,
    "props": Object {
      "fileName": Object {
        "end": 46,
        "start": 31,
        "value": "Tess",
        "valueEnd": 45,
        "valueStart": 41,
      },
      "key": Object {
        "end": 30,
        "start": 20,
        "value": "TESS",
        "valueEnd": 29,
        "valueStart": 25,
      },
      "player": Object {
        "end": 62,
        "start": 47,
        "value": "TonyLB",
        "valueEnd": 61,
        "valueStart": 55,
      },
    },
    "start": 9,
    "tag": "Character",
    "tagEnd": 19,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should correctly subset by property 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "contents": Array [
          Object {
            "end": 150,
            "start": 117,
            "type": "string",
            "value": "Test Render:
                    ",
          },
          Object {
            "contents": Array [
              Object {
                "end": 194,
                "start": 183,
                "type": "string",
                "value": "Clock Tower",
              },
            ],
            "contentsEnd": 194,
            "contentsStart": 183,
            "end": 201,
            "props": Object {
              "key": Object {
                "end": 166,
                "start": 156,
                "value": "a123",
                "valueEnd": 165,
                "valueStart": 161,
              },
              "to": Object {
                "end": 182,
                "start": 167,
                "value": "clockTower",
                "valueEnd": 181,
                "valueStart": 171,
              },
            },
            "start": 150,
            "tag": "Link",
            "tagEnd": 155,
            "type": "tag",
          },
        ],
        "contentsEnd": 218,
        "contentsStart": 96,
        "end": 232,
        "props": Object {},
        "start": 83,
        "tag": "Description",
        "tagEnd": 95,
        "type": "tag",
      },
      Object {
        "contents": Array [
          Object {
            "end": 269,
            "start": 265,
            "type": "string",
            "value": "test",
          },
        ],
        "contentsEnd": 269,
        "contentsStart": 265,
        "end": 276,
        "props": Object {
          "to": Object {
            "end": 264,
            "start": 255,
            "value": "Test",
            "valueEnd": 263,
            "valueStart": 259,
          },
        },
        "start": 249,
        "tag": "Exit",
        "tagEnd": 254,
        "type": "tag",
      },
      Object {
        "contents": Array [
          Object {
            "end": 317,
            "start": 311,
            "type": "string",
            "value": "vortex",
          },
        ],
        "contentsEnd": 317,
        "contentsStart": 311,
        "end": 324,
        "props": Object {
          "from": Object {
            "end": 310,
            "start": 299,
            "value": "Test",
            "valueEnd": 309,
            "valueStart": 305,
          },
        },
        "start": 293,
        "tag": "Exit",
        "tagEnd": 298,
        "type": "tag",
      },
    ],
    "contentsEnd": 337,
    "contentsStart": 66,
    "end": 344,
    "props": Object {
      "global": Object {
        "end": 65,
        "start": 59,
        "value": true,
      },
      "key": Object {
        "end": 58,
        "start": 46,
        "value": "VORTEX",
        "valueEnd": 57,
        "valueStart": 51,
      },
    },
    "start": 40,
    "tag": "Room",
    "tagEnd": 45,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should properly chain complex predicates 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "end": 269,
        "start": 265,
        "type": "string",
        "value": "test",
      },
    ],
    "contentsEnd": 269,
    "contentsStart": 265,
    "end": 276,
    "props": Object {
      "to": Object {
        "end": 264,
        "start": 255,
        "value": "Test",
        "valueEnd": 263,
        "valueStart": 259,
      },
    },
    "start": 249,
    "tag": "Exit",
    "tagEnd": 254,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should properly select :first filter 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "end": 269,
        "start": 265,
        "type": "string",
        "value": "test",
      },
    ],
    "contentsEnd": 269,
    "contentsStart": 265,
    "end": 276,
    "props": Object {
      "to": Object {
        "end": 264,
        "start": 255,
        "value": "Test",
        "valueEnd": 263,
        "valueStart": 259,
      },
    },
    "start": 249,
    "tag": "Exit",
    "tagEnd": 254,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should properly select boolean or 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "end": 150,
        "start": 117,
        "type": "string",
        "value": "Test Render:
                    ",
      },
      Object {
        "contents": Array [
          Object {
            "end": 194,
            "start": 183,
            "type": "string",
            "value": "Clock Tower",
          },
        ],
        "contentsEnd": 194,
        "contentsStart": 183,
        "end": 201,
        "props": Object {
          "key": Object {
            "end": 166,
            "start": 156,
            "value": "a123",
            "valueEnd": 165,
            "valueStart": 161,
          },
          "to": Object {
            "end": 182,
            "start": 167,
            "value": "clockTower",
            "valueEnd": 181,
            "valueStart": 171,
          },
        },
        "start": 150,
        "tag": "Link",
        "tagEnd": 155,
        "type": "tag",
      },
    ],
    "contentsEnd": 218,
    "contentsStart": 96,
    "end": 232,
    "props": Object {},
    "start": 83,
    "tag": "Description",
    "tagEnd": 95,
    "type": "tag",
  },
  Object {
    "contents": Array [
      Object {
        "end": 556,
        "start": 466,
        "type": "string",
        "value": "Clocktower
                    test
                    on multiple lines
                ",
      },
    ],
    "contentsEnd": 556,
    "contentsStart": 445,
    "end": 570,
    "props": Object {},
    "start": 432,
    "tag": "Description",
    "tagEnd": 444,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should properly select boolean or of chained selectors 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "end": 269,
        "start": 265,
        "type": "string",
        "value": "test",
      },
    ],
    "contentsEnd": 269,
    "contentsStart": 265,
    "end": 276,
    "props": Object {
      "to": Object {
        "end": 264,
        "start": 255,
        "value": "Test",
        "valueEnd": 263,
        "valueStart": 259,
      },
    },
    "start": 249,
    "tag": "Exit",
    "tagEnd": 254,
    "type": "tag",
  },
  Object {
    "contents": Array [
      Object {
        "end": 317,
        "start": 311,
        "type": "string",
        "value": "vortex",
      },
    ],
    "contentsEnd": 317,
    "contentsStart": 311,
    "end": 324,
    "props": Object {
      "from": Object {
        "end": 310,
        "start": 299,
        "value": "Test",
        "valueEnd": 309,
        "valueStart": 305,
      },
    },
    "start": 293,
    "tag": "Exit",
    "tagEnd": 298,
    "type": "tag",
  },
  Object {
    "contents": Array [
      Object {
        "end": 556,
        "start": 466,
        "type": "string",
        "value": "Clocktower
                    test
                    on multiple lines
                ",
      },
    ],
    "contentsEnd": 556,
    "contentsStart": 445,
    "end": 570,
    "props": Object {},
    "start": 432,
    "tag": "Description",
    "tagEnd": 444,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should properly select grouped operators 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "end": 150,
        "start": 117,
        "type": "string",
        "value": "Test Render:
                    ",
      },
      Object {
        "contents": Array [
          Object {
            "end": 194,
            "start": 183,
            "type": "string",
            "value": "Clock Tower",
          },
        ],
        "contentsEnd": 194,
        "contentsStart": 183,
        "end": 201,
        "props": Object {
          "key": Object {
            "end": 166,
            "start": 156,
            "value": "a123",
            "valueEnd": 165,
            "valueStart": 161,
          },
          "to": Object {
            "end": 182,
            "start": 167,
            "value": "clockTower",
            "valueEnd": 181,
            "valueStart": 171,
          },
        },
        "start": 150,
        "tag": "Link",
        "tagEnd": 155,
        "type": "tag",
      },
    ],
    "contentsEnd": 218,
    "contentsStart": 96,
    "end": 232,
    "props": Object {},
    "start": 83,
    "tag": "Description",
    "tagEnd": 95,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should properly select nested :nthChild filters 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "end": 317,
        "start": 311,
        "type": "string",
        "value": "vortex",
      },
    ],
    "contentsEnd": 317,
    "contentsStart": 311,
    "end": 324,
    "props": Object {
      "from": Object {
        "end": 310,
        "start": 299,
        "value": "Test",
        "valueEnd": 309,
        "valueStart": 305,
      },
    },
    "start": 293,
    "tag": "Exit",
    "tagEnd": 298,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should properly select one level :nthChild filter 1`] = `
Array [
  Object {
    "contents": Array [],
    "end": 376,
    "props": Object {
      "key": Object {
        "end": 373,
        "start": 363,
        "value": "Test",
        "valueEnd": 372,
        "valueStart": 368,
      },
    },
    "start": 357,
    "tag": "Room",
    "tagEnd": 362,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should return a list for multiple matches 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "contents": Array [
          Object {
            "end": 150,
            "start": 117,
            "type": "string",
            "value": "Test Render:
                    ",
          },
          Object {
            "contents": Array [
              Object {
                "end": 194,
                "start": 183,
                "type": "string",
                "value": "Clock Tower",
              },
            ],
            "contentsEnd": 194,
            "contentsStart": 183,
            "end": 201,
            "props": Object {
              "key": Object {
                "end": 166,
                "start": 156,
                "value": "a123",
                "valueEnd": 165,
                "valueStart": 161,
              },
              "to": Object {
                "end": 182,
                "start": 167,
                "value": "clockTower",
                "valueEnd": 181,
                "valueStart": 171,
              },
            },
            "start": 150,
            "tag": "Link",
            "tagEnd": 155,
            "type": "tag",
          },
        ],
        "contentsEnd": 218,
        "contentsStart": 96,
        "end": 232,
        "props": Object {},
        "start": 83,
        "tag": "Description",
        "tagEnd": 95,
        "type": "tag",
      },
      Object {
        "contents": Array [
          Object {
            "end": 269,
            "start": 265,
            "type": "string",
            "value": "test",
          },
        ],
        "contentsEnd": 269,
        "contentsStart": 265,
        "end": 276,
        "props": Object {
          "to": Object {
            "end": 264,
            "start": 255,
            "value": "Test",
            "valueEnd": 263,
            "valueStart": 259,
          },
        },
        "start": 249,
        "tag": "Exit",
        "tagEnd": 254,
        "type": "tag",
      },
      Object {
        "contents": Array [
          Object {
            "end": 317,
            "start": 311,
            "type": "string",
            "value": "vortex",
          },
        ],
        "contentsEnd": 317,
        "contentsStart": 311,
        "end": 324,
        "props": Object {
          "from": Object {
            "end": 310,
            "start": 299,
            "value": "Test",
            "valueEnd": 309,
            "valueStart": 305,
          },
        },
        "start": 293,
        "tag": "Exit",
        "tagEnd": 298,
        "type": "tag",
      },
    ],
    "contentsEnd": 337,
    "contentsStart": 66,
    "end": 344,
    "props": Object {
      "global": Object {
        "end": 65,
        "start": 59,
        "value": true,
      },
      "key": Object {
        "end": 58,
        "start": 46,
        "value": "VORTEX",
        "valueEnd": 57,
        "valueStart": 51,
      },
    },
    "start": 40,
    "tag": "Room",
    "tagEnd": 45,
    "type": "tag",
  },
  Object {
    "contents": Array [],
    "end": 376,
    "props": Object {
      "key": Object {
        "end": 373,
        "start": 363,
        "value": "Test",
        "valueEnd": 372,
        "valueStart": 368,
      },
    },
    "start": 357,
    "tag": "Room",
    "tagEnd": 362,
    "type": "tag",
  },
]
`;

exports[`wmlQuery selector should select root node when passed empty string 1`] = `
Array [
  Object {
    "contents": Array [
      Object {
        "contents": Array [
          Object {
            "end": 86,
            "start": 82,
            "type": "string",
            "value": "Tess",
          },
        ],
        "contentsEnd": 86,
        "contentsStart": 82,
        "end": 93,
        "props": Object {},
        "start": 76,
        "tag": "Name",
        "tagEnd": 81,
        "type": "tag",
      },
      Object {
        "contents": Array [],
        "contentsEnd": 290,
        "contentsStart": 290,
        "end": 301,
        "props": Object {
          "adjective": Object {
            "end": 240,
            "start": 224,
            "value": "hers",
            "valueEnd": 239,
            "valueStart": 235,
          },
          "object": Object {
            "end": 174,
            "start": 162,
            "value": "her",
            "valueEnd": 173,
            "valueStart": 170,
          },
          "possessive": Object {
            "end": 207,
            "start": 191,
            "value": "her",
            "valueEnd": 206,
            "valueStart": 203,
          },
          "reflexive": Object {
            "end": 276,
            "start": 257,
            "value": "herself",
            "valueEnd": 275,
            "valueStart": 268,
          },
          "subject": Object {
            "end": 145,
            "start": 132,
            "value": "she",
            "valueEnd": 144,
            "valueStart": 141,
          },
        },
        "start": 106,
        "tag": "Pronouns",
        "tagEnd": 115,
        "type": "tag",
      },
      Object {
        "contents": Array [
          Object {
            "end": 342,
            "start": 331,
            "type": "string",
            "value": "Frumpy Goth",
          },
        ],
        "contentsEnd": 342,
        "contentsStart": 331,
        "end": 360,
        "props": Object {},
        "start": 314,
        "tag": "FirstImpression",
        "tagEnd": 330,
        "type": "tag",
      },
      Object {
        "contents": Array [
          Object {
            "end": 399,
            "start": 387,
            "type": "string",
            "value": "Fuchsia eyes",
          },
        ],
        "contentsEnd": 399,
        "contentsStart": 387,
        "end": 414,
        "props": Object {},
        "start": 373,
        "tag": "OneCoolThing",
        "tagEnd": 386,
        "type": "tag",
      },
      Object {
        "contents": Array [
          Object {
            "end": 517,
            "start": 435,
            "type": "string",
            "value": "A bulky frock-coat lovingly kit-bashed from a black hoodie and patchily dyed lace.",
          },
        ],
        "contentsEnd": 517,
        "contentsStart": 435,
        "end": 526,
        "props": Object {},
        "start": 427,
        "tag": "Outfit",
        "tagEnd": 434,
        "type": "tag",
      },
    ],
    "contentsEnd": 535,
    "contentsStart": 63,
    "end": 547,
    "props": Object {
      "fileName": Object {
        "end": 46,
        "start": 31,
        "value": "Tess",
        "valueEnd": 45,
        "valueStart": 41,
      },
      "key": Object {
        "end": 30,
        "start": 20,
        "value": "TESS",
        "valueEnd": 29,
        "valueStart": 25,
      },
      "player": Object {
        "end": 62,
        "start": 47,
        "value": "TonyLB",
        "valueEnd": 61,
        "valueStart": 55,
      },
    },
    "start": 9,
    "tag": "Character",
    "tagEnd": 19,
    "type": "tag",
  },
]
`;
